internal/stores/hostdb.go

```diff
@@ -365,64 +365,95 @@ func (db *SQLStore) HostBlocklist() (blocklist []string, err error) {
 
 // RecordHostInteraction records an interaction with a host. If the host is not in
 // the store, a new entry is created for it.
-func (db *SQLStore) RecordHostInteractions(hostKey consensus.PublicKey, interactions []hostdb.Interaction) error {
-	dbInteractions := make([]dbInteraction, len(interactions))
-	var successful, failed float64
-
-	var scanQuery string
-	var scanArgs []interface{}
-	for i, interaction := range interactions {
-		dbInteractions[i] = dbInteraction{
-			Result:    interaction.Result,
-			Success:   interaction.Success,
-			Timestamp: interaction.Timestamp.UTC(),
-			Type:      interaction.Type,
+func (db *SQLStore) RecordHostInteractions(interactions []hostdb.Interaction) error {
+	// Get keys from input.
+	keyMap := make(map[consensus.PublicKey]struct{})
+	for _, interaction := range interactions {
+		keyMap[interaction.Host] = struct{}{}
+	}
+	var hks [][]byte
+	for hk := range keyMap {
+		hks = append(hks, gobEncode(hk))
+	}
+
+	return db.db.Transaction(func(tx *gorm.DB) error {
+		// Fetch hosts for which to add interactions.
+		var hosts []dbHost
+		if err := db.db.Where("public_key IN ?", hks).
+			Find(&hosts).Error; err != nil {
+			return err
 		}
-		if interaction.Success {
-			successful++
-		} else {
-			failed++
+		hostMap := make(map[consensus.PublicKey]dbHost)
+		for _, h := range hosts {
+			hostMap[h.PublicKey] = h
 		}
-		// Handle scan edge case.
-		if interaction.Type == hostdb.InteractionTypeScan {
-			timeToAddTo := "uptime"
-			if !interaction.Success {
-				timeToAddTo = "downtime"
+
+		// Apply all the interactions to the hosts.
+		dbInteractions := make([]dbInteraction, len(interactions))
+		for i, interaction := range interactions {
+			isScan := interaction.Type == hostdb.InteractionTypeScan
+			dbInteractions[i] = dbInteraction{
+				Result:    interaction.Result,
+				Success:   interaction.Success,
+				Timestamp: interaction.Timestamp.UTC(),
```

Comment:
maybe we should consider getting rid of all of our timestamps since we found out we need timestamps to do math on them?

Reply:
Yeah but probably in a separate PR. I only want to touch what we really need in this one.